File: test/e2e/setup_test.go
// go:build e2e

package e2e_test

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/harness/harness-mcp/cmd/harness-mcp-server/config"
	"github.com/harness/harness-mcp/pkg/harness"
	mcpClient "github.com/mark3labs/mcp-go/client"
	"github.com/mark3labs/mcp-go/mcp"
	"github.com/stretchr/testify/require"
)

var (
	// Shared variables and sync.Once instances to ensure one-time execution
	getTokenOnce sync.Once
	token        string
	tokenErr     error

	getAccountIDOnce sync.Once
	accountID        string
	accountIDErr     error

	getOrgIDOnce sync.Once
	orgID        string

	getProjectIDOnce sync.Once
	projectID        string

	setupClientOnce sync.Once
	testClient      *mcpClient.Client
	setupClientErr  error
)

// getE2EToken ensures the environment variable is checked only once and returns the token
func getE2EToken(t *testing.T) (string, error) {
	getTokenOnce.Do(func() {
		token = os.Getenv("HARNESS_MCP_SERVER_E2E_TOKEN")
		if token == "" {
			tokenErr = fmt.Errorf("HARNESS_MCP_SERVER_E2E_TOKEN environment variable is not set")
		}
	})
	if tokenErr != nil {
		return "", tokenErr
	}
	return token, nil
}

// getE2EAccountID ensures the environment variable is checked only once and returns the account ID
func getE2EAccountID(t *testing.T) (string, error) {
	getAccountIDOnce.Do(func() {
		// First check if explicitly set
		accountID = os.Getenv("HARNESS_MCP_SERVER_E2E_ACCOUNT_ID")
		if accountID != "" {
			return
		}

		// If not set, try to extract from PAT token
		pat, err := getE2EToken(t)
		if err != nil {
			accountIDErr = err
			return
		}
		// PAT format is pat.{account_id}.{token_id}.{token_value}
		parts := strings.Split(pat, ".")
		if len(parts) >= 2 {
			accountID = parts[1]
			return
		}

		// Try to get from HARNESS_MCP_USER_PAT if HARNESS_MCP_SERVER_E2E_TOKEN not set
		pat = os.Getenv("HARNESS_MCP_USER_PAT")
		if pat != "" {
			parts := strings.Split(pat, ".")
			if len(parts) >= 2 {
				accountID = parts[1]
				return
			}
		}

		accountIDErr = fmt.Errorf("could not determine account ID from token or environment variables")
	})
	if accountIDErr != nil {
		return "", accountIDErr
	}
	return accountID, nil
}

// getE2EOrgID ensures the environment variable is checked only once and returns the org ID
func getE2EOrgID() string {
	getOrgIDOnce.Do(func() {
		orgID = os.Getenv("HARNESS_MCP_SERVER_E2E_ORG_ID")
		if orgID == "" {
			// Default to a test org ID if not specified
			orgID = "AI_Devops"
		}
	})
	return orgID
}

// getE2EProjectID ensures the environment variable is checked only once and returns the project ID
func getE2EProjectID() string {
	getProjectIDOnce.Do(func() {
		projectID = os.Getenv("HARNESS_MCP_SERVER_E2E_PROJECT_ID")
		if projectID == "" {
			// Default to a test project ID if not specified
			projectID = "Sanity"
		}
	})
	return projectID
}

// clientOpts holds configuration options for the MCP client setup
type clientOpts struct {
	// Toolsets to enable in the MCP server
	enabledToolsets []string
}

// clientOption defines a function type for configuring ClientOpts
type clientOption func(*clientOpts)

// withToolsets returns an option that sets the toolsets to enable
func withToolsets(toolsets []string) clientOption {
	return func(opts *clientOpts) {
		opts.enabledToolsets = toolsets
	}
}

// setupMCPClient creates and initializes an MCP client for testing
func setupMCPClient(t *testing.T, options ...clientOption) *mcpClient.Client {
	// Only setup once across all tests
	setupClientOnce.Do(func() {
		// Get token
		token, err := getE2EToken(t)
		if err != nil {
			setupClientErr = err
			return
		}
		accountID, err := getE2EAccountID(t)
		if err != nil {
			setupClientErr = err
			return
		}
		// Create and configure options
		opts := &clientOpts{
			enabledToolsets: []string{"ccm", "pipelines", "default"},
		}

		// Apply all options to configure the opts struct
		for _, option := range options {
			option(opts)
		}

		// Create a basic config for the MCP server
		cfg := &config.Config{
			APIKey:           token,
			Toolsets:         opts.enabledToolsets,
			EnableLicense:    false,
			ReadOnly:         true,
			AccountID:        accountID,
			DefaultOrgID:     getE2EOrgID(),
			DefaultProjectID: getE2EProjectID(),
			BaseURL:          os.Getenv("HARNESS_MCP_SERVER_E2E_BASE_URL"),
		}

		// Initialize toolsets
		tsg, err := harness.InitToolsets(context.Background(), cfg)
		if err != nil {
			setupClientErr = err
			return
		}

		// Create an MCP server instance with default options
		mcpServer := harness.NewServer("0.0.1")

		// Register the tools with the server
		tsg.RegisterTools(mcpServer)

		// Create an in-process MCP client
		testClient, err = mcpClient.NewInProcessClient(mcpServer)
		if err != nil {
			setupClientErr = err
			return
		}

		// Initialize the client
		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
		defer cancel()

		request := mcp.InitializeRequest{}
		request.Params.ProtocolVersion = "2025-03-26"
		request.Params.ClientInfo = mcp.Implementation{
			Name:    "harness-e2e-test-client",
			Version: "0.0.1",
		}

		_, err = testClient.Initialize(ctx, request)
		if err != nil {
			setupClientErr = err
			return
		}
	})

	// Check if setup was successful
	require.NoError(t, setupClientErr, "expected to setup MCP client successfully")

	return testClient
}

// ensureDockerImageBuilt makes sure the Docker image is built only once across all tests
func ensureDockerImageBuilt(t *testing.T) {
	buildOnce := sync.Once{}
	var buildError error

	buildOnce.Do(func() {
		t.Log("Building Docker image for e2e tests...")
		cmd := exec.Command("docker", "build", "-t", "harness/e2e-mcp-server", ".")
		output, err := cmd.CombinedOutput()
		buildError = err
		if err != nil {
			t.Logf("Docker build output: %s", string(output))
		}
	})

	// Check if the build was successful
	require.NoError(t, buildError, "expected to build Docker image successfully")
}

// runE2ETest is a helper function to run a test with the MCP client
func runE2ETest(t *testing.T, testFunc func(t *testing.T, client *mcpClient.Client, ctx context.Context)) {
	t.Helper()

	// Setup the MCP client
	client := setupMCPClient(t)

	// Create a context with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	// Run the test function
	testFunc(t, client, ctx)
}

File: tests/e2e/setup_test.rs
use crate::config::Config;
use crate::harness;
use crate::harness::prompts;
use crate::modules;
use std::env;
use std::sync::Once;
use tokio::sync::OnceCell;

// Shared variables and Once instances to ensure one-time execution
static GET_TOKEN_ONCE: Once = Once::new();
static mut TOKEN: Option<String> = None;

static GET_ACCOUNT_ID_ONCE: Once = Once::new();
static mut ACCOUNT_ID: Option<String> = None;

static GET_ORG_ID_ONCE: Once = Once::new();
static mut ORG_ID: Option<String> = None;

static GET_PROJECT_ID_ONCE: Once = Once::new();
static mut PROJECT_ID: Option<String> = None;

static SETUP_CLIENT_ONCE: OnceCell<Result<TestClient, Box<dyn std::error::Error + Send + Sync>>> = OnceCell::const_new();

#[derive(Clone)]
pub struct TestClient {
    // MCP client would go here when available
    // For now, we'll use a placeholder
    pub base_url: String,
    pub token: String,
}

/// Get E2E token, ensures the environment variable is checked only once
pub fn get_e2e_token() -> Result<String, Box<dyn std::error::Error>> {
    unsafe {
        GET_TOKEN_ONCE.call_once(|| {
            TOKEN = env::var("HARNESS_MCP_SERVER_E2E_TOKEN").ok();
        });
        
        TOKEN.clone().ok_or_else(|| {
            "HARNESS_MCP_SERVER_E2E_TOKEN environment variable is not set".into()
        })
    }
}

/// Get E2E account ID, ensures the environment variable is checked only once
pub fn get_e2e_account_id() -> Result<String, Box<dyn std::error::Error>> {
    unsafe {
        GET_ACCOUNT_ID_ONCE.call_once(|| {
            ACCOUNT_ID = env::var("HARNESS_MCP_SERVER_E2E_ACCOUNT_ID").ok();
        });
        
        ACCOUNT_ID.clone().ok_or_else(|| {
            "HARNESS_MCP_SERVER_E2E_ACCOUNT_ID environment variable is not set".into()
        })
    }
}

/// Get E2E org ID, ensures the environment variable is checked only once
pub fn get_e2e_org_id() -> Result<String, Box<dyn std::error::Error>> {
    unsafe {
        GET_ORG_ID_ONCE.call_once(|| {
            ORG_ID = env::var("HARNESS_MCP_SERVER_E2E_ORG_ID").ok();
        });
        
        ORG_ID.clone().ok_or_else(|| {
            "HARNESS_MCP_SERVER_E2E_ORG_ID environment variable is not set".into()
        })
    }
}

/// Get E2E project ID, ensures the environment variable is checked only once
pub fn get_e2e_project_id() -> Result<String, Box<dyn std::error::Error>> {
    unsafe {
        GET_PROJECT_ID_ONCE.call_once(|| {
            PROJECT_ID = env::var("HARNESS_MCP_SERVER_E2E_PROJECT_ID").ok();
        });
        
        PROJECT_ID.clone().ok_or_else(|| {
            "HARNESS_MCP_SERVER_E2E_PROJECT_ID environment variable is not set".into()
        })
    }
}

/// Setup test client, ensures client is created only once
pub async fn setup_test_client() -> Result<TestClient, Box<dyn std::error::Error + Send + Sync>> {
    SETUP_CLIENT_ONCE.get_or_init(|| async {
        let token = get_e2e_token()?;
        
        // Create test client
        let client = TestClient {
            base_url: "https://app.harness.io".to_string(),
            token,
        };
        
        Ok(client)
    }).await.clone()
}

/// Helper function to skip test if environment variables are not set
pub fn skip_if_no_e2e_env() -> Result<(), Box<dyn std::error::Error>> {
    get_e2e_token()?;
    get_e2e_account_id()?;
    get_e2e_org_id()?;
    get_e2e_project_id()?;
    Ok(())
}

/// Create test config for E2E tests
pub fn create_test_config() -> Result<Config, Box<dyn std::error::Error>> {
    let token = get_e2e_token()?;
    let account_id = get_e2e_account_id()?;
    let org_id = get_e2e_org_id()?;
    let project_id = get_e2e_project_id()?;
    
    Ok(Config {
        harness_platform_api_key: Some(token),
        harness_platform_account_id: Some(account_id),
        harness_platform_organization_id: Some(org_id),
        harness_platform_project_id: Some(project_id),
        internal: false,
        ..Default::default()
    })
}

/// Helper to run command and capture output
pub async fn run_command(cmd: &str, args: &[&str]) -> Result<String, Box<dyn std::error::Error>> {
    let output = tokio::process::Command::new(cmd)
        .args(args)
        .output()
        .await?;
    
    if !output.status.success() {
        let stderr = String::from_utf8_lossy(&output.stderr);
        return Err(format!("Command failed: {}", stderr).into());
    }
    
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

/// Helper to check if binary exists
pub async fn check_binary_exists(binary_name: &str) -> bool {
    tokio::process::Command::new("which")
        .arg(binary_name)
        .output()
        .await
        .map(|output| output.status.success())
        .unwrap_or(false)
}

#[cfg(test)]
mod tests {
    use super::*;
    use tokio;

    #[tokio::test]
    async fn test_get_e2e_token() {
        // This test will only pass if the environment variable is set
        if env::var("HARNESS_MCP_SERVER_E2E_TOKEN").is_ok() {
            let token = get_e2e_token().expect("Should get token");
            assert!(!token.is_empty());
        }
    }

    #[tokio::test]
    async fn test_skip_if_no_e2e_env() {
        // This test checks if all required environment variables are set
        // It will skip if they're not available
        match skip_if_no_e2e_env() {
            Ok(_) => {
                // All environment variables are set
                println!("E2E environment variables are available");
            }
            Err(_) => {
                // Some environment variables are missing
                println!("E2E environment variables are not set, skipping E2E tests");
            }
        }
    }

    #[tokio::test]
    async fn test_create_test_config() {
        if skip_if_no_e2e_env().is_ok() {
            let config = create_test_config().expect("Should create config");
            assert!(config.harness_platform_api_key.is_some());
            assert!(config.harness_platform_account_id.is_some());
            assert!(config.harness_platform_organization_id.is_some());
            assert!(config.harness_platform_project_id.is_some());
        }
    }

    #[tokio::test]
    async fn test_setup_test_client() {
        if skip_if_no_e2e_env().is_ok() {
            let client = setup_test_client().await.expect("Should setup client");
            assert!(!client.token.is_empty());
            assert!(!client.base_url.is_empty());
        }
    }

    #[tokio::test]
    async fn test_run_command() {
        let output = run_command("echo", &["hello", "world"])
            .await
            .expect("Should run echo command");
        assert!(output.contains("hello world"));
    }

    #[tokio::test]
    async fn test_check_binary_exists() {
        // Test with a binary that should exist on most systems
        let exists = check_binary_exists("echo").await;
        assert!(exists);
        
        // Test with a binary that shouldn't exist
        let not_exists = check_binary_exists("nonexistent_binary_12345").await;
        assert!(!not_exists);
    }
}